runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkDefaultExecutableStageContext.java

```diff
@@ -17,18 +17,34 @@
  */
 package org.apache.beam.runners.flink.translation.functions;
 
+import com.google.common.collect.ImmutableMap;
+import org.apache.beam.model.pipeline.v1.RunnerApi.StandardEnvironments;
+import org.apache.beam.runners.core.construction.BeamUrns;
+import org.apache.beam.runners.core.construction.Environments;
 import org.apache.beam.runners.core.construction.graph.ExecutableStage;
-import org.apache.beam.runners.fnexecution.control.DockerJobBundleFactory;
 import org.apache.beam.runners.fnexecution.control.JobBundleFactory;
+import org.apache.beam.runners.fnexecution.control.LazyJobBundleFactory;
 import org.apache.beam.runners.fnexecution.control.StageBundleFactory;
+import org.apache.beam.runners.fnexecution.environment.DockerEnvironmentFactory;
+import org.apache.beam.runners.fnexecution.environment.InProcessEnvironmentFactory;
+import org.apache.beam.runners.fnexecution.environment.ProcessEnvironmentFactory;
 import org.apache.beam.runners.fnexecution.provisioning.JobInfo;
 
 /** Implementation of a {@link FlinkExecutableStageContext}. */
 class FlinkDefaultExecutableStageContext implements FlinkExecutableStageContext, AutoCloseable {
   private final JobBundleFactory jobBundleFactory;
 
   private static FlinkDefaultExecutableStageContext create(JobInfo jobInfo) throws Exception {
-    JobBundleFactory jobBundleFactory = DockerJobBundleFactory.FACTORY.get().create(jobInfo);
+    JobBundleFactory jobBundleFactory =
+        LazyJobBundleFactory.create(
+            jobInfo,
+            ImmutableMap.of(
+                BeamUrns.getUrn(StandardEnvironments.Environments.DOCKER),
+                new DockerEnvironmentFactory.Provider(),
+                BeamUrns.getUrn(StandardEnvironments.Environments.PROCESS),
+                new ProcessEnvironmentFactory.Provider(),
+                Environments.ENVIRONMENT_EMBEDDED, // Non Public urn for testing.
+                new InProcessEnvironmentFactory.Provider()));
```

Comment:
rename factory

Reply:
Done

Reply:
I wonder if this could be made easier to customize / pluggable. For Lyft, we want to override the factory provider for the default environment for Python. It would be nice if we can just drop an annotated class into the classpath, similar to what we now do for native transforms. This can be outside of this PR, but it would be good to support: https://issues.apache.org/jira/browse/BEAM-4819 @mxm

Reply:
I agree, I was thinking of a ServiceRegistry but decided to keep it as a separate effort to reduce complexity of the PR.